File: My progarmming/machine-learning-ex6/ex6_matlab/ex6_spam.m
%% Machine Learning Online Class
%  Exercise 6 | Spam Classification with SVMs
%
%  Instructions
%  ------------
% 
%  This file contains code that helps you get started on the
%  exercise. You will need to complete the following functions:
%
%     gaussianKernel.m
%     dataset3Params.m
%     processEmail.m
%     emailFeatures.m
%
%  For this exercise, you will not need to change any code in this file,
%  or any other files other than those mentioned above.
%

%% Initialization
clear ; close all; clc

%% ==================== Part 1: Email Preprocessing ====================
%  To use an SVM to classify emails into Spam v.s. Non-Spam, you first need
%  to convert each email into a vector of features. In this part, you will
%  implement the preprocessing steps for each email. You should
%  complete the code in processEmail.m to produce a word indices vector
%  for a given email.

fprintf('\nPreprocessing sample email (emailSample1.txt)\n');

% Extract Features
file_contents = readFile('emailSample1.txt');%readFile()读取文件并返回其全部内容
word_indices  = processEmail(file_contents);%processEmail()预处理电子邮件正文并返回单词索引列表

% Print Stats
fprintf('Word Indices: \n');
fprintf(' %d', word_indices);
fprintf('\n\n');

fprintf('Program paused. Press enter to continue.\n');
pause;

%% ==================== Part 2: Feature Extraction ====================
%  Now, you will convert each email into a vector of features in R^n. 
%  You should complete the code in emailFeatures.m to produce a feature
%  vector for a given email.

fprintf('\nExtracting features from sample email (emailSample1.txt)\n');

% Extract Features
file_contents = readFile('emailSample1.txt');
word_indices  = processEmail(file_contents);
features      = emailFeatures(word_indices);%emailFeatures()接受单词索引向量并从单词索引生成特征向量

% Print Stats
fprintf('Length of feature vector: %d\n', length(features));
fprintf('Number of non-zero entries: %d\n', sum(features > 0));

fprintf('Program paused. Press enter to continue.\n');
pause;

%% =========== Part 3: Train Linear SVM for Spam Classification ========
%  In this section, you will train a linear classifier to determine if an
%  email is Spam or Not-Spam.

% Load the Spam Email dataset
% You will have X, y in your environment
load('spamTrain.mat');

fprintf('\nTraining Linear SVM (Spam Classification)\n')
fprintf('(this may take 1 to 2 minutes) ...\n')

C = 0.1;
model = svmTrain(X, y, C, @linearKernel);%svmTrain()训练支持向量机分类器并返回训练模型;linearKernel()返回x1和x2之间的线性核函数

p = svmPredict(model, X);%使用经过训练的SVM模型（svmTrain）返回预测向量

fprintf('Training Accuracy: %f\n', mean(double(p == y)) * 100);

%% =================== Part 4: Test Spam Classification ================
%  After training the classifier, we can evaluate it on a test set. We have
%  included a test set in spamTest.mat

% Load the test dataset
% You will have Xtest, ytest in your environment
load('spamTest.mat');

fprintf('\nEvaluating the trained Linear SVM on a test set ...\n')

p = svmPredict(model, Xtest);

fprintf('Test Accuracy: %f\n', mean(double(p == ytest)) * 100);

fprintf('\nProgram paused. Press enter to continue.\n');
pause;


%% ================= Part 5: Top Predictors of Spam ====================
%  Since the model we are training is a linear SVM, we can inspect the
%  weights learned by the model to understand better how it is determining
%  whether an email is spam or not. The following code finds the words with
%  the highest weights in the classifier. Informally, the classifier
%  'thinks' that these words are the most likely indicators of spam.
%

% Sort the weights and obtin the vocabulary list
[weight, idx] = sort(model.w, 'descend');%mode为'descend'时，进行降序排序，weight是排序好的向量，idx是向量weight中对model.w的索引
vocabList = getVocabList();

fprintf('\nTop predictors of spam: \n');
for i = 1:15
    fprintf(' %-15s (%f) \n', vocabList{idx(i)}, weight(i));
end

fprintf('\n\n');
fprintf('\nProgram paused. Press enter to continue.\n');
pause;

%% =================== Part 6: Try Your Own Emails =====================
%  Now that you've trained the spam classifier, you can use it on your own
%  emails! In the starter code, we have included spamSample1.txt,
%  spamSample2.txt, emailSample1.txt and emailSample2.txt as examples. 
%  The following code reads in one of these emails and then uses your 
%  learned SVM classifier to determine whether the email is Spam or 
%  Not Spam

% Set the file to be read in (change this to spamSample2.txt,
% emailSample1.txt or emailSample2.txt to see different predictions on
% different emails types). Try your own emails as well!
filename = 'emailSample1.txt';
% filename = 'emailSample2.txt';

% Read and predict
file_contents = readFile(filename);
word_indices  = processEmail(file_contents);
x             = emailFeatures(word_indices);
p = svmPredict(model, x);

fprintf('\nProcessed %s\n\nSpam Classification: %d\n', filename, p);
fprintf('(1 indicates spam, 0 indicates not spam)\n\n');

File: My progarmming/machine-learning-ex6/ex6_python/ex6_spam.py
import matplotlib.pyplot as plt
import numpy as np
import scipy.io as scio
from sklearn import svm
import re#处理正则表达式的模块
import nltk, nltk.stem.porter
 
plt.ion()#使matplotlib的显示模式转换为交互(interactive)模式
np.set_printoptions(formatter={'float': '{: 0.6f}'.format})
 
'''================================================part1 邮件预处理============================================='''

def process_email(email_contents):
    vocab_list = get_vocab_list()#vocab_list为1899x1
    word_indices = np.array([], dtype=np.int64)
 
    # ===================== Preprocess Email =====================
 
    email_contents = email_contents.lower()
    email_contents = re.sub('<[^<>]+>', ' ', email_contents)
    email_contents = re.sub('[0-9]+', 'number', email_contents)
    email_contents = re.sub('(http|https)://[^\s]*', 'httpaddr', email_contents)
    email_contents = re.sub('[^\s]+@[^\s]+', 'emailaddr', email_contents)
    email_contents = re.sub('[$]+', 'dollar', email_contents)
 
    # ===================== Tokenize Email =====================
 
    print('==== Processed Email ====')
    stemmer = nltk.stem.porter.PorterStemmer()
    tokens = re.split('[@$/#.-:&*+=\[\]?!(){\},\'\">_<;% ]', email_contents)
 
    for token in tokens:
        token = re.sub('[^a-zA-Z0-9]', '', token)
        token = stemmer.stem(token)
 
        if len(token) < 1:
            continue
 
        for i in range(1, len(vocab_list) + 1):
            if vocab_list[i] == token:
                word_indices = np.append(word_indices, i)

        print(token)
 
    print('==================')
 
    return word_indices
 
def get_vocab_list(): #得到词汇列表
    vocab_dict = {}     #新建空字典 并以字典形式获取
    with open('vocab.txt') as f:
        for line in f:
            (val, key) = line.split()  #读取每一行的键和值
            vocab_dict[int(val)] = key #存放到字典中
 
    return vocab_dict

# ===================== Part 1: Email Preprocessing =====================
 
print('Preprocessing sample email (emailSample1.txt) ...')
 
file_contents = open('emailSample1.txt', 'r').read()
word_indices = process_email(file_contents)
 
# Print stats
print('Word Indices: ')
print(word_indices)
 
input('Program paused. Press ENTER to continue')
 
'''================================================part2 提取特征============================================='''

def email_features(word_indices):
    # Total number of words in the dictionary
    n = 1899
 
    # Since the index of numpy array starts at 0, to align with the word indices we make n + 1 size array
    features = np.zeros((n + 1,1))#(1-1899在python中从0开始，要得到1899需要1900大小的元组)
 
    for i in word_indices:
        features[i] =  1
 
    return features

# ===================== Part 2: Feature Extraction =====================
 
print('Extracting Features from sample email (emailSample1.txt) ... ')
 
# Extract features
features = email_features(word_indices)
 
# Print stats
print('Length of feature vector: {}'.format(features.size))
print('Number of non-zero entries: {}'.format(np.flatnonzero(features).size))
 
input('Program paused. Press ENTER to continue')

'''================================================part3 训练SVM============================================='''

# ===================== Part 3: Train Linear SVM for Spam Classification =====================
 
data = scio.loadmat('spamTrain.mat')
X = data['X']
y = data['y'].flatten()
 
print('Training Linear SVM (Spam Classification)')
print('(this may take 1 to 2 minutes)')
 
c = 0.1
clf = svm.SVC(c, kernel='linear')
clf.fit(X, y)
 
p = clf.predict(X)#X 4000x1899
 
print('Training Accuracy: {}'.format(np.mean(p == y) * 100))

# ===================== Part 4: Test Spam Classification =====================

# Load the test dataset
data = scio.loadmat('spamTest.mat')
Xtest = data['Xtest']
ytest = data['ytest'].flatten()
 
print('Evaluating the trained linear SVM on a test set ...')
 
p = clf.predict(Xtest)#Xtest 1000x1899
 
print('Test Accuracy: {}'.format(np.mean(p == ytest) * 100))
 
input('Program paused. Press ENTER to continue')

'''================================================part4 高权重词============================================='''

# ===================== Part 5: Top Predictors of Spam =====================
 
vocab_list = get_vocab_list()
indices = np.argsort(clf.coef_).flatten()[::-1]
print(indices)
 
for i in range(15):
    print('{} ({:0.6f})'.format(vocab_list[indices[i]], clf.coef_.flatten()[indices[i]]))
 
input('Program paused. Press ENTER to continue')

# =================== Part 6: Try Your Own Emails =====================

filename = 'spamSample1.txt';
file_contents = open(filename, 'r').read()
word_indices = process_email(file_contents)
x = email_features(word_indices)#x 1900x1
p = clf.predict(x[1:,:].T)#x[1:,:].T 1X1899

print('\nProcessed ', filename,'\nSpam Classification:', p);
print('(1 indicates spam, 0 indicates not spam)\n\n');

input('ex6_spam Finished. Press ENTER to exit')
